hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/sort/SpaceCurveSortingHelper.java

```diff
@@ -205,48 +179,91 @@ public boolean hasNext() {
-            } else if (dataType instanceof DateType) {
-              return row.isNullAt(index) ? Long.MAX_VALUE : row.getDate(index).getTime();
-            } else if (dataType instanceof TimestampType) {
-              return row.isNullAt(index) ? Long.MAX_VALUE : row.getTimestamp(index).getTime();
-            } else if (dataType instanceof ByteType) {
-              return row.isNullAt(index) ? Long.MAX_VALUE : BinaryUtil.convertBytesToLong(new byte[] {row.getByte(index)});
-            } else if (dataType instanceof ShortType) {
-              return row.isNullAt(index) ? Long.MAX_VALUE : (long)row.getShort(index);
-            } else if (dataType instanceof DecimalType) {
-              return row.isNullAt(index) ? Long.MAX_VALUE : row.getDecimal(index).longValue();
-            } else if (dataType instanceof BooleanType) {
-              boolean value = row.isNullAt(index) ? false : row.getBoolean(index);
-              return value ? Long.MAX_VALUE : 0;
-            } else if (dataType instanceof BinaryType) {
-              return row.isNullAt(index) ? Long.MAX_VALUE : BinaryUtil.convertBytesToLong((byte[]) row.get(index));
-            }
-            return null;
-          }).filter(f -> f != null).collect(Collectors.toList());
+          long[] longs = fieldMap.entrySet().stream()
+              .mapToLong(entry -> {
+                int index = entry.getKey();
+                StructField field = entry.getValue();
+                return mapColumnValueToLong(row, index, field.dataType());
+              })
+              .toArray();
 
-          byte[] hilbertValue = HilbertCurveUtils.indexBytes(
-              hilbertCurve, longList.stream().mapToLong(l -> l).toArray(), 63);
-          List<Object> values = new ArrayList<>();
-          values.addAll(scala.collection.JavaConverters.bufferAsJavaListConverter(row.toSeq().toBuffer()).asJava());
-          values.add(hilbertValue);
-          return Row$.MODULE$.apply(JavaConversions.asScalaBuffer(values));
+          // Map N-dimensional coordinates into position on the Hilbert curve
+          byte[] hilbertCurvePosBytes = HilbertCurveUtils.indexBytes(hilbertCurve, longs, 63);
+          return appendToRow(row, hilbertCurvePosBytes);
         }
       };
-    }).sortBy(f -> new ByteArraySorting((byte[]) f.get(fieldNum)), true, fileNum);
+    })
+        .sortBy(f -> new ByteArraySorting((byte[]) f.get(fieldNum)), true, fileNum);
+  }
+
+  private static Row appendToRow(Row row, Object value) {
+    // NOTE: This is an ugly hack to avoid array re-allocation --
+    //       Spark's {@code Row#toSeq} returns array of Objects
+    Object[] currentValues = (Object[]) ((WrappedArray<Object>) row.toSeq()).array();
+    return RowFactory.create(CollectionUtils.append(currentValues, value));
+  }
+
+  @Nonnull
+  private static byte[] mapColumnValueTo8Bytes(Row row, int index, DataType dataType) {
```

Comment:
this is very hard to review. and pushes all the overhead to the reviewer.  Can you comment on the PR to call out what has actually changed?  

> please separate refactoring with the actual implementation of functionality

I'd also prefer if we did these restructuring/moving in a separate PR when possible. makes for an easier review.

Reply:
Yeah, this commit was incorrectly placed into this branch instead of the one it was stacked on.

Summary of the changes here:
1. This large conditional was extracted to standalone method (no changes w/in conditional itself)
2. In the surrounding context following things change
2.a Added `appendToRow` method to avoid double conversion from Scala > Java > Scala
2.b Use primitive Stream methods (`mapToLong`) to avoid unnecessary boxing